src/charm.py

```diff
@@ -8,35 +8,208 @@
+            return
+        self.certificates_provider.request_certificate_revocation(
+            event.certificate_signing_request.encode()
+        )
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        """Handles certificate available events.
+
+        Finds the relation ID matching the CSR and forwards the received
+        certificate to that relation.
+        If a relation ID is not found, logs an error and ignores the event.
+
+        Args:
+            event (CertificateAvailableEvent): Event containing the certificate
+
+        Returns:
+            None
+        """
+        relation_id = self._get_relation_id_for_csr(event.certificate_signing_request)
+        if not relation_id:
+            logger.error(
+                "Could not find the relation for CSR: %s.",
+                event.certificate_signing_request,
+            )
+            return
+        self.certificates_requirers.set_relation_certificate(
+            certificate=event.certificate,
+            certificate_signing_request=event.certificate_signing_request,
+            ca=event.ca,
+            chain=event.chain,
+            relation_id=relation_id,
+        )
+
+    def _on_certificate_invalidated(self, event: CertificateInvalidatedEvent) -> None:
+        """Handles certificate invalidated events.
+
+        If the certificate is invalidated because it expired, ignore the event
+        and let the requirer handle it properly. Otherwise, calls the TLS
+        library to revoke the certificate to the requirer.
+
+        Args:
+            event (CertificateInvalidatedEvent): Event for invalidated certificate
+
+        Returns:
+            None
+        """
+        if event.reason == "expired":
+            return
+        self.certificates_requirers.remove_certificate(event.certificate)
+
+    def _on_all_certificates_invalidated(self, event: AllCertificatesInvalidatedEvent) -> None:
+        """Handles all certificates invalidated events.
+
+        Revokes all certificates.
+
+        Args:
+            event (AllCertificatesInvalidatedEvent): Event for all certificates invalidated
+
+        Returns:
+            None
+        """
+        self.certificates_requirers.revoke_all_certificates()
+
+    def _get_relation_id_for_csr(self, csr: str) -> Optional[int]:
+        """Finds the relation ID that sent the provided CSR.
+
+        This should return a single relation ID, otherwise it means multiple
+        applications requested the same certificate using the same private
+        key. In that case, we log an error and return None.
+
+        Args:
+            csr (str): Certificate Signing Request to search
+
+        Returns:
+            Relation ID (int) or None
+        """
+        requirers_csrs = self.certificates_requirers.get_requirer_csrs()
+        relation_ids: set[int] = set()
+        for requirer_csrs in requirers_csrs:
+            match requirer_csrs:
```

Comment:
Maybe a bit clearer? I missed the second s 5 times before I noticed (python allows using the same variable names because one is inside the for loop scope too for some reason, so if the variable names were the same it would still work).
```suggestion
        for requirer_csrs in all_requirers_csrs:
            match requirer_csrs:
```

Reply:
Agreed, this would make things clearer.

Reply:
Fixed